doc/api/advanced_input.rst

```diff
@@ -0,0 +1,91 @@
+Advanced Input
+==============
+
+.. automodule:: simpleline.input.input_handler
+
+.. WARNING::
+    This section contains advanced input techniques which may lead to a buggy code in your
+    program if they are not used correctly.
+
+The default technique how to get user input is already described in
+:ref:`UIScreen <uiscreen_label>` and that should be a preferred way how to obtain a user input.
+However, there are reasons when those can't be used. For those reasons it is possible to go deeper
+low level in the input obtaining and implement a custom solution.
+
+Input handler classes
+---------------------
+Input handler classes are classes created to obtain a user input. If required
+:class:`InputHandler` can block application until a user input is received.
+
+This class can be instantiated everywhere in the code and used to ask for user input::
+
+    handler = InputHandler()
+    handler.get_input("Shut up and give me your input:")
+    handler.wait_for_input()
+    if handler.input_successful:
+        user_input = handler.value
+
+The :meth:`InputHandler.wait_on_input` method will block code processing until a user input is
+received. The input should be always checked if successful before processing.
+
+In the case some other work needs to be done before a user input is received then pass
+a callback to the constructor of the :class:`InputHandler` class and do not use
+the :meth:`InputHandler.wait_on_input` method. However, the callback is using
+the :ref:`event loop<event_loops_label>` so the callback won't be called until the event loop
+processing is active. If :ref:`concurrent input<concurrent_input_label>` is used then this
+callback may never be called!
```

Comment:
"may never be called" -- do you mean it must never be called, or it might never get called? "May" has a very specific definition, but people tend to use it in place of a lot of other words, and it's confusing here.